### PR TITLE
Update codebase to use Three.js version r88

### PR DESCRIPTION
**Pull request by jihoonl**

### CODE REVIEW

**Comment by jihoonl** [comment]
Well.. three@0.64.0 doesn't exist. The closest version is 0.66.0. Check if it is compatible.

**Comment by jihoonl** [comment]
Well.. let's bump the codebase to r88 instead of sticking with old version.

**Comment by jihoonl** [comment]
Hello @viktorku @sevenbitbyte , I have updated codebase to work with three.js r88. Please check if you have a time.

**Comment by jihoonl** [comment]
I have confirmed that marker, interactive marker, and urdf demos are working with this change.

File: build/ros3d.js [hunk 1]
```diff
@@ -791,8 +790,8 @@ ROS3D.InteractiveMarker.prototype.onServerSetPose = function(event) {
       this.position.y = pose.position.y;
       this.position.z = pose.position.z;
 
-      this.quaternion = new THREE.Quaternion(pose.orientation.x, pose.orientation.y,
-          pose.orientation.z, pose.orientation.w);
+      this.quaternion.copy(new THREE.Quaternion(pose.orientation.x, pose.orientation.y,
```
Comment (T045T) [inline build/ros3d.js]: you could probably use set() here, as you did with the other quaternions further down

File: src/visualization/interaction/Highlighter.js [hunk 2]
```diff
@@ -35,62 +36,43 @@ ROS3D.Highlighter.prototype.onMouseOver = function(event) {
...
-  var objlist = scene.__webglObjects;
-  // get corresponding webgl objects
-  for ( var c = 0; c < objects.length; c++) {
-    if (objects[c]) {
-      for ( var o = objlist.length - 1; o >= 0; o--) {
-        if (objlist[o].object === objects[c]) {
-          renderList.push(objlist[o]);
-          break;
-        }
+ROS3D.Highlighter.prototype.highlightObject = function (object, flag) {
```
Comment (T045T) [inline src/visualization/interaction/Highlighter.js]: This doesn't actually highlight correctly, but instead replaces the highlighted objects' materials with a half-transparent white material:
screen shot 2017-12-11 at 15 04 47

The old implementation rendered that material *over* the already-rendered scene to achieve the highlighted look. 

Looks like current Three.js has a couple ways to make highlighting happen:
[This demo](https://threejs.org/examples/?q=out#webgl_postprocessing_outline) ([source](https://github.com/mrdoob/three.js/blob/master/examples/webgl_postprocessing_outline.html)) uses an ```OutlinePass``` to draw outlines, which looks really nice but is slightly more involved ([here](https://stackoverflow.com/questions/26341396/outline-a-3d-object-in-three-js)'s a StackOverflow question with some very concise example code).

[This one](https://threejs.org/examples/#webgl_interactive_cubes) ([source](https://github.com/mrdoob/three.js/blob/master/examples/webgl_interactive_cubes.html)) simply gives the material a bit of (red) ```emissive``` lighting to achieve about the same effect we were using before (note they use ```currentHex``` to save the old value – I'd prefer to use ```originalHex```).

**Comment by jihoonl** [comment]
Hello @DLu, @jstnhuang, and @rctoris,

I am trying to update the codebase to the latest three.js(r88) and found out that fetch isn't loadable due to [this error in three.js](https://github.com/mrdoob/three.js/issues/12855). Could you help me to fix this issue? I have confirmed that I can visualize our robot, turtlebot2, and turtlebot3 with no problem.

**Comment by jihoonl** [comment]
Another thing for @DLu and @jstnhuang,

I had to revert #187 to properly visualize turtlebot3 and pr2. Could you check if it breaks your visualization? Also, somehow pr2 shows its collision model instead of a proper mesh.. Do you know how to fix this mesh visualization issue?

With #187 
pr2_with_187
turtlebot3

**After reverting #187**
pr2_without_187
without_187

File: src/visualization/interaction/Highlighter.js (hunk 2 above)
Comment (jihoonl) [reply src/visualization/interaction/Highlighter.js]: The second example is the one I referenced to update the highlighting. Thanks for the review. I will take a look again.

**Comment by jstnhuang** [comment]
Hi Jihoon, here are my notes:

First, getting things set up:
- I know bower is deprecated, but could we bump the version number for three.js in bower.json, as well?
- The copies of ColladaLoader.js and STLLoader.js I had saved didn't work with the new version of THREE.js. Instead, I copied [ColladaLoader.js](https://github.com/mrdoob/three.js/blob/dev/examples/js/loaders/ColladaLoader.js) and [STLLoader.js](https://github.com/mrdoob/three.js/blob/dev/examples/js/loaders/STLLoader.js) from the dev branch of three.js

With this setup, I tried visualizing the PR2, Fetch, Turtlebot 2, and Baxter:

PR2:
pr2

Fetch:
fetch

Turtlebot 2:
turtlebot

Baxter:
baxter

Unfortunately, I don't know what's causing these problems. Here's what I see in the console:

PR2: No errors or warnings.

Fetch: Multiple errors like this:
```
Cannot read property 'build' of undefined (ColladaLoader.js:207)
    at getBuild (ColladaLoader.js:207)
    at getCamera (ColladaLoader.js:1627)
    at buildNode (ColladaLoader.js:3040)
    at getBuild (ColladaLoader.js:209)
    at getNode (ColladaLoader.js:3235)
    at buildVisualScene (ColladaLoader.js:3281)
    at buildLibrary (ColladaLoader.js:197)
    at THREE.ColladaLoader.parse (ColladaLoader.js:3424)
    at Object.onLoad (ColladaLoader.js:27)
    at XMLHttpRequest.<anonymous> (three.js:30483)
```

Turtlebot 2: Many warnings like this:
```
three.js:18506 THREE.WebGLRenderer: image is not power of two (600x600). Resized to 512x512 
<img crossorigin="Anonymous" src="http://localhost:8001/turtlebot_description/meshes/stacks/hexagons/./images/3f_pole.jpg">
```

Baxter: No errors or warnings. It's strange that part of the robot has the correct TF and some does not.

Sorry if this isn't helpful. I might be able to try and look into it myself later.

**Comment by jihoonl** [comment]
@jstnhuang FYI, you may want to check out https://github.com/mrdoob/three.js/issues/12855 to understand the problem.

**Comment by jihoonl** [comment]
ANd this https://github.com/mrdoob/three.js/pull/12856

**Comment by T045T** [comment]
Since RViz (presumably) loads all of these models fine, does that imply it handles the ```up_axis``` information in the DAE files incorrectly, or do we need to take a closer look at how ros3djs handles ColladaLoader's output?

**Comment by T045T** [comment]
Update: I checked, and RViz really does ignore the ```up_axis``` tag in DAE files (see https://github.com/ros-visualization/rviz/issues/1045 ).

Since ColladaLoader doesn't have an option to disable proper handling of ```up_axis```, I see two options here:

1. Petition RViz to change their behavior, potentially breaking existing ```description``` packages (like the one for Baxter)
2. Fork ColladaLoader and include the forked version (possibly directly in ros3d.js ?)

I think option 2, while less clean, is more realistic.

**Comment by jihoonl** [comment]
Thanks for the investigation. I was bugging this in mrdoob/three.js#12856.

**Comment by jihoonl** [comment]
Error occurring while loading fetch urdf will be fixed with [this PR in three.js](https://github.com/mrdoob/three.js/pull/12866). But, It infers that r88 isn't the good target to use as codbase of ros3djs. We may want to wait until next release or find older version that works. I think the next release will come before February.

@T045T I agree with you that option2 is less clean but more realistic. Let's go for option2.

**Comment by jihoonl** [comment]
Note that this up_axis issue appear in all `kobuki`, `turtlebot`, `fetch`, and `baxter` descriptions.

**Comment by T045T** [comment]
I went ahead and sent some quick PRs to those repos - but of course, we should still be compatible with RViz' behavior.

For future reference, the command I used to automatically remove all up_axis tags was

    find . -name *.dae | xargs sed -i '/up_axis/d'

in the top-level workspace directory.

File: src/visualization/interaction/Highlighter.js [hunk 3]
```diff
@@ -35,62 +35,42 @@ ROS3D.Highlighter.prototype.onMouseOver = function(event) {
...
+        opacity : 0.5,
+        depthTest : true,
+        depthWrite : false,
+        polygonOffset : true,
+        polygonOffsetUnits : -1,
+        side : THREE.DoubleSide
+      });
+    }
+    else {
+      target_object.material = target_object.originalMaterial;
```
Comment (T045T) [inline src/visualization/interaction/Highlighter.js]: One more thing: We should probably be defensive here and check for the existence of ```originalMaterial```.

If ```highlightObject(false)``` is ever called on an object where ```highlight(true)``` has not been called, we'd be in trouble!

**Comment by jihoonl** [comment]
**Summary**
*  the forked collada loader is in examples/js/ColladaLoader2.js. It includes z_axis hack to avoid [rviz issue that ignoring up_axis](https://github.com/ros-visualization/rviz/issues/1045), and cherry-picked [this commit](https://github.com/mrdoob/three.js/commit/14e9d73fe176d32387337d3c5b188d45eff6aa49#diff-844b1cc1e411ca1bec3c6b4115a2af4c)
* three.js version r88
* Highlighter logic is fied to compatibly with r88. It still needs some more improvements. This will be fixed with the following PRs.
* Note that browser doesn't support `tif` format. This needs to be converted to `png` format.
```
#!/bin/bash

for f in $(find -name *.tif); do
  echo "Converting $f"; convert "$f"  "$(dirname "$f")/$(basename "$f" .tif).png";
done

for f in $(find -name *.dae); do
  echo "Replacing $f"; sed -i -- 's/.tif/.png/g' $f
done
```

File: src/visualization/interaction/Highlighter.js (hunk 3 above)
Comment (jihoonl) [reply src/visualization/interaction/Highlighter.js]: Yeah, I agree. I would improve this logic with another PR.

Comment (T045T) [reply src/visualization/interaction/Highlighter.js]: Please don't. Merging this PR as-is would introduce a serious bug (see the screen shot I attached to the issue), and fixing it would be trivial (just give the existing material an emissive component instead of switching the material completely).

File: examples/js/ColladaLoader2.js [hunk 4]
```diff
@@ -0,0 +1,3476 @@
...
+ *
+ * @Modified by Jihoon Lee from ColladerLoader.js@r88
+ * To support rviz compatible collada viewing.
+ * See: #202 why it is forked.
+ *
+ * It is a fork from ColladerLoader.js in three.js. It follows three.js license.
+ * Also it is renamed as ColladaLoader2.js
+ */
+
+THREE.ColladaLoader2 = function ( manager ) {
```
Comment (T045T) [inline examples/js/ColladaLoader2.js]: Why did you name it ColladaLoader2? It's ColladaLoader, and we actually want it to replace THREE.ColladaLoader completely, don't we?

On a higher level, do we even still need to support two loaders? Looks like what was once ColladaLoader2 has replaced the original ColladaLoader completely by now anyway.

File: examples/map.html [hunk 5]
```diff
@@ -3,9 +3,9 @@
 <head>
 <meta charset="utf-8" />
 
-<script src="http://cdn.robotwebtools.org/threejs/current/three.js"></script>
-<script src="http://cdn.robotwebtools.org/EventEmitter2/current/eventemitter2.min.js"></script>
-<script src="http://cdn.robotwebtools.org/roslibjs/current/roslib.js"></script>
+<script src="http://static.robotwebtools.org/threejs/current/three.js"></script>
```
Comment (T045T) [inline examples/map.html]: these changes should really go in a separate PR - the CDN has already migrated, and we should reflect that in our code ASAP

Comment (jihoonl) [reply examples/map.html]: It has been already requested to @rctoris. And it doesn't matter as long as it doesn't go in to master branch.

File: examples/js/ColladaLoader2.js (hunk 4 above)
Comment (jihoonl) [reply examples/js/ColladaLoader2.js]: Why do we need to replace it completely? It needs to be explicitly kept separate so that it indicates it is a modified version just for backward compatibility. Users should know that they use the modified version because their collada isn't following the standard.

Comment (T045T) [reply examples/js/ColladaLoader2.js]: I disagree. 

My opinion is that users shouldn't have to care about what loader is used at all – de facto, we're not dealing with the Collada standard, but instead with the one set by RViz. From that perspective, all models that work in RViz are correct and we should display them the same way RViz does.

Either way, renaming the module to ColladaLoader2 is confusing because that's not what it is.  It has the exact same interface as ```THREE.ColladaLoader```.

Because of that, we could actually remove all the ```loader``` options from ```ros3djs``` and instead just use a different script tag to load either the official ColladaLoader version or this patched one:

```html
    <!-- Uncomment this whenever RViz starts handling up_axis
    <script src="http://static.robotwebtools.org/threejs/current/ColladaLoader.js"></script>
    -->
    <!-- Remove this when RViz starts handling up_axis -->
    <script src="http://static.robotwebtools.org/fixes/ColladaLoader.js"></script>
```

File: examples/map.html (hunk 5 above)
Comment (T045T) [reply examples/map.html]: I'm not sure I understand what you're saying. The new URLs (with static in them) already work. 

Why don't we make a separate pull request to update them in examples and Readme files right now, while we're still discussing the move to ThreeJS r88?

File: src/visualization/interaction/Highlighter.js (hunk 3 above)
Comment (jihoonl) [reply src/visualization/interaction/Highlighter.js]: It is not as easy as you described and it doesn't introduce serious bug while I was testing. This PR is already pretty large and getting longer and I am running out of time. Since it is already a big change which covers most of use cases. I will move on and start to patch small stuffs. I don't see any serious reason that this should not go into the develop branch.

File: examples/js/ColladaLoader2.js (hunk 4 above)
Comment (jihoonl) [reply examples/js/ColladaLoader2.js]: Ok. reasonable. Also, I removed all loader option because it doesn't seem necessary.

File: examples/map.html (hunk 5 above)
Comment (jihoonl) [reply examples/map.html]: What I am saying is that our stable branch is `master` where we should keep the working code. I am moving on. We need to create a PR to fix some links example anyway when CDN is updated.

**Comment by jihoonl** [comment]
@T045T sorry Nils. I proceeded to merge because it was a definitely needed movement. Otherwise, this PR would hang for another year considering the recent year's activities in this repo.
